src/main/java/de/fearnixx/jeak/teamspeak/voice/connection/VoiceConnection.java

```diff
@@ -0,0 +1,223 @@
+                        onError.run();
+                        return;
+                    }
+
+                    this.connected = true;
+                    onSuccess.run();
+                }
+        ).start();
+    }
+
+    @Override
+    public void disconnect() {
+        disconnect(null);
+    }
+
+    @Override
+    public void disconnect(String reason) {
+        try {
+            if (ts3jClientSocket.getMicrophone() != null
+                    && ts3jClientSocket.getMicrophone().getClass().isAssignableFrom(IAudioPlayer.class)) {
+                ((IAudioPlayer) ts3jClientSocket.getMicrophone()).stop();
+            }
+
+            ts3jClientSocket.disconnect(reason);
+        } catch (InterruptedException ie) {
+            Thread.currentThread().interrupt();
+        } catch (IOException | ExecutionException | TimeoutException e) {
+            //We assume that the server is not reachable for this connection. Therefor it is declared disconnected
+        }
+
+        locked = false;
+        connected = false;
+    }
+
+    @Override
+    public boolean isConnected() {
+        return connected;
+    }
+
+    @Override
+    public void sendToChannel(int channelId) {
+        sendToChannel(channelId, null);
+    }
+
+    @Override
+    public void sendToChannel(int channelId, String password) {
+        try {
+            ts3jClientSocket.clientMove(ts3jClientSocket.getClientId(), channelId, password);
+        } catch (InterruptedException ie) {
+            Thread.currentThread().interrupt();
+        } catch (IOException | TimeoutException | CommandException e) {
+            //The client movement failed
+
+            //TODO: return the result of an action performed on the connection
+        }
+    }
+
+    @Override
+    public IAudioPlayer registerAudioPlayer(AudioType audioType) {
+        if (!connected) {
+            throw new IllegalStateException("An audio player can only be registered when the client connection is connected");
+        }
+
+        AudioPlayer audioPlayer;
+
+        switch (audioType) {
+            case MP3:
+                audioPlayer = new Mp3AudioPlayer();
+                break;
+            default:
+                throw new IllegalArgumentException("The audio type is currently not supported!");
+        }
+
+        ts3jClientSocket.setMicrophone(audioPlayer);
+
+        return audioPlayer;
+    }
+
+    @Override
+    public Optional<IAudioPlayer> optRegisteredAudioPlayer() {
+        if (ts3jClientSocket == null || ts3jClientSocket.getMicrophone() == null) {
+            return Optional.empty();
+        } else {
+            return Optional.of((IAudioPlayer) ts3jClientSocket.getMicrophone());
+        }
+    }
+
+    @Override
+    public IAudioPlayer getRegisteredAudioPlayer() {
+        return optRegisteredAudioPlayer().orElseThrow(
+                () -> new IllegalStateException("The voice connection had no registered audio player!")
```

Comment:
Please include the voice connections identifier so administrators have a clue more.